In [65]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder,OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
import numpy as np

In [71]:
dt=pd.read_csv('50_Startups.csv')
dt.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [72]:
datax=dt.iloc[:,0:4].values
datax
datay=dt.iloc[:,-1].values

In [73]:
le=LabelEncoder()

In [74]:
datax[:,3]=le.fit_transform(datax[:,3]) #encodes the data at column number 4

In [75]:
datax

array([[165349.2, 136897.8, 471784.1, 2],
       [162597.7, 151377.59, 443898.53, 0],
       [153441.51, 101145.55, 407934.54, 1],
       [144372.41, 118671.85, 383199.62, 2],
       [142107.34, 91391.77, 366168.42, 1],
       [131876.9, 99814.71, 362861.36, 2],
       [134615.46, 147198.87, 127716.82, 0],
       [130298.13, 145530.06, 323876.68, 1],
       [120542.52, 148718.95, 311613.29, 2],
       [123334.88, 108679.17, 304981.62, 0],
       [101913.08, 110594.11, 229160.95, 1],
       [100671.96, 91790.61, 249744.55, 0],
       [93863.75, 127320.38, 249839.44, 1],
       [91992.39, 135495.07, 252664.93, 0],
       [119943.24, 156547.42, 256512.92, 1],
       [114523.61, 122616.84, 261776.23, 2],
       [78013.11, 121597.55, 264346.06, 0],
       [94657.16, 145077.58, 282574.31, 2],
       [91749.16, 114175.79, 294919.57, 1],
       [86419.7, 153514.11, 0.0, 2],
       [76253.86, 113867.3, 298664.47, 0],
       [78389.47, 153773.43, 299737.29, 2],
       [73994.56, 122782.75, 30331

In [76]:
onehe=OneHotEncoder(categorical_features=[3]) 

In [77]:
datax=onehe.fit_transform(datax).toarray() #transforms the the whole data

/home/usman/.local/lib/python3.6/site-packages/sklearn/preprocessing/_encoders.py:415: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)


In [78]:
datax[0:5]

array([[0.0000000e+00, 0.0000000e+00, 1.0000000e+00, 1.6534920e+05,
        1.3689780e+05, 4.7178410e+05],
       [1.0000000e+00, 0.0000000e+00, 0.0000000e+00, 1.6259770e+05,
        1.5137759e+05, 4.4389853e+05],
       [0.0000000e+00, 1.0000000e+00, 0.0000000e+00, 1.5344151e+05,
        1.0114555e+05, 4.0793454e+05],
       [0.0000000e+00, 0.0000000e+00, 1.0000000e+00, 1.4437241e+05,
        1.1867185e+05, 3.8319962e+05],
       [0.0000000e+00, 1.0000000e+00, 0.0000000e+00, 1.4210734e+05,
        9.1391770e+04, 3.6616842e+05]])

In [79]:
mle=LinearRegression() #building the model

In [80]:
trainx,testx,trainy,testy=train_test_split(datax,datay,test_size=0.2,random_state=0)
#splitting the model into training and testing dataset

In [81]:
mle.fit(trainx,trainy) #training the model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [82]:
predict_y=mle.predict(testx) #giving the testing data, to check it's prediction

In [83]:
predict_y 

array([103015.20159795, 132582.27760816, 132447.73845175,  71976.09851258,
       178537.48221057, 116161.24230167,  67851.69209676,  98791.73374687,
       113969.43533014, 167921.06569552])

In [84]:
testy

array([103282.38, 144259.4 , 146121.95,  77798.83, 191050.39, 105008.31,
        81229.06,  97483.56, 110352.25, 166187.94])

In [88]:
prediction_difference=[]
for i,k in zip(predict_y,testy):
    prediction_difference.append(i-k)
#calculating the dfference between predicted and test output

In [89]:
prediction_difference=np.array(prediction_difference) #converting the difference into an array

In [90]:
prediction_difference

array([  -267.17840205, -11677.12239184, -13674.21154825,  -5822.73148742,
       -12512.90778943,  11152.93230167, -13377.36790324,   1308.17374687,
         3617.18533014,   1733.12569552])

# the above value is the difference between the predicted and the test output
# As you can see that at points there is alot of difference between like at index number 5 which is 11152.93 which is a huge difference , so we can say that the performance of this Simple MultiLinear Regression is not good at all

In [91]:
predict_y-testy

array([  -267.17840205, -11677.12239184, -13674.21154825,  -5822.73148742,
       -12512.90778943,  11152.93230167, -13377.36790324,   1308.17374687,
         3617.18533014,   1733.12569552])